crates/iroha_core/src/state.rs

```diff
@@ -1100,7 +1082,7 @@ impl State {
     }
 
     /// Create structure to execute a block
-    pub fn block(&self) -> StateBlock<'_> {
+    pub fn block(&self, curr_block: BlockHeader) -> StateBlock<'_> {
```

Comment:
I don't think this can take any args because the transition to `StateBlock` should be unique. What if let `State` hold the last committed block header which `StateBlock` inherit and mutate to the voting block header as needed?

Reply:
I see your point. I believe I hinted at this in the PR description. I think this could be done as an optimization in a separate PR. For the moment it has to be that `block` takes the header of the block currently processed

Reply:
> let `State` hold the last committed block header

Or call `StateReadOnly::latest_block` on the transition to initialize a field specific to `StateBlock` and `StateTransaction` with the last committed block header

But yeah, for now developers involved in this logic can be careful not to pass an arbitrary block header

Reply:
created #5162